From b066da81c4a548b3cf6a561ba786eea7be60aa50 Mon Sep 17 00:00:00 2001
From: Simeon Marijon <simeon.marijon@foss.st.com>
Date: Mon, 11 Dec 2023 12:48:15 +0100
Subject: [PATCH 434/954] feat(st-nvmem): add nvmem framework

Based on linux and U-boot API, allows users to access
to nvmem based on nvmem cell names

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 5a37a82baffb0775b0aa0ae48cc8a28ba13701e4 feat(st-nvmem): add nvmem framework]
Signed-off-by: Simeon Marijon <simeon.marijon@foss.st.com>
Change-Id: I1a1641770d50cb424391d03606268a96e57f8e09
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/344691
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Simeon MARIJON <simeon.marijon@st.com>
Reviewed-by: Simeon MARIJON <simeon.marijon@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 changelog.yaml                  |   3 +
 drivers/st/nvmem/nvmem.c        | 133 ++++++++++++++++++++++++++++++++
 include/drivers/st/nvmem.h      |  45 +++++++++++
 plat/st/stm32mp1/stm32mp1_def.h |   3 +
 plat/st/stm32mp2/stm32mp2_def.h |   3 +
 5 files changed, 187 insertions(+)
 create mode 100644 drivers/st/nvmem/nvmem.c
 create mode 100644 include/drivers/st/nvmem.h

diff --git a/changelog.yaml b/changelog.yaml
index 77cdea8b5a..89c9be719e 100644
--- a/changelog.yaml
+++ b/changelog.yaml
@@ -1141,6 +1141,9 @@ subsections:
           - title: GPIO
             scope: st-gpio
 
+          - title: NVMEM
+            scope: st-nvmem
+
           - title: Regulator
             scope: st-regulator
 
diff --git a/drivers/st/nvmem/nvmem.c b/drivers/st/nvmem/nvmem.c
new file mode 100644
index 0000000000..00688aeb7d
--- /dev/null
+++ b/drivers/st/nvmem/nvmem.c
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2023, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <drivers/st/nvmem.h>
+#include <lib/libfdt/libfdt.h>
+#include <lib/libfdt/libfdt_env.h>
+
+#include <platform_def.h>
+
+#define for_each_nvmem_dev(nvmem_dev) \
+	for ((nvmem_dev) = nvmem_dev_array; \
+	     (nvmem_dev) <= &nvmem_dev_array[PLAT_NB_NVMEM_DEVS - 1U]; \
+	     (nvmem_dev)++)
+
+static struct nvmem_dev nvmem_dev_array[PLAT_NB_NVMEM_DEVS];
+
+int nvmem_driver_register(int nodeoffset, const struct nvmem_ops *nvmem_ops,
+			  void *drv_data)
+{
+	struct nvmem_dev *nvmem_dev = NULL;
+
+	for_each_nvmem_dev(nvmem_dev) {
+		if (nvmem_dev->drv_data == NULL) {
+			break;
+		}
+	}
+
+	if (nvmem_dev > &nvmem_dev_array[PLAT_NB_NVMEM_DEVS - 1U]) {
+		WARN("Not enough place for NVMEM dev, PLAT_NB_NVMEM_DEVS should be increased.\n");
+		return -ENOMEM;
+	}
+
+	nvmem_dev->nvmem_ops = nvmem_ops;
+	nvmem_dev->drv_data = drv_data;
+	nvmem_dev->offset = nodeoffset;
+
+	return 0;
+}
+
+static struct nvmem_dev *nvmem_get_by_node(int node_offset)
+{
+	struct nvmem_dev *nvmem_dev = NULL;
+
+	for_each_nvmem_dev(nvmem_dev) {
+		if (nvmem_dev->offset == node_offset) {
+			return nvmem_dev;
+		}
+	}
+
+	WARN("%s: node %d not found\n", __func__, node_offset);
+	return NULL;
+}
+
+int nvmem_get_cell_by_index(const void *fdt, int nodeoffset, unsigned int index,
+			    struct nvmem_cell *cell)
+{
+	int len = 0;
+	const fdt32_t *nvmem_cells_prop = NULL, *nvmem_cell_reg_prop = NULL;
+	uint32_t phandle;
+	int nvmem_cell_node;
+	int parent_node;
+	struct nvmem_dev *dev = NULL;
+
+	nvmem_cells_prop = fdt_getprop(fdt, nodeoffset, "nvmem-cells", &len);
+	if (nvmem_cells_prop == NULL) {
+		ERROR("Can't find nvmem cell");
+		return len;
+	}
+
+	phandle = fdt32_to_cpu(nvmem_cells_prop[index]);
+	nvmem_cell_node = fdt_node_offset_by_phandle(fdt, phandle);
+
+	nvmem_cell_reg_prop = fdt_getprop(fdt, nvmem_cell_node, "reg", &len);
+
+	if ((nvmem_cell_reg_prop == NULL) ||
+	    ((unsigned int)len != (sizeof(uint32_t) * U(2)))) {
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	cell->offset = fdt32_to_cpu(nvmem_cell_reg_prop[0]);
+	cell->len = fdt32_to_cpu(nvmem_cell_reg_prop[1]);
+
+	parent_node = fdt_parent_offset(fdt, nvmem_cell_node);
+	dev = nvmem_get_by_node(parent_node);
+	if (dev == NULL) {
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	cell->dev = dev;
+	return 0;
+}
+
+int nvmem_get_cell_by_name(const void *fdt, int nodeoffset, const char *name,
+			   struct nvmem_cell *cell)
+{
+	int index = fdt_stringlist_search(fdt, nodeoffset, "nvmem-cell-names", name);
+
+	if (index < 0) {
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	return nvmem_get_cell_by_index(fdt, nodeoffset, (unsigned int)index,
+				       cell);
+};
+
+int nvmem_cell_read(struct nvmem_cell *cell, uint8_t *output_buf,
+		    size_t buf_len, size_t *read_len)
+{
+	if (cell->dev->nvmem_ops->read_cell == NULL) {
+		return -ENOTSUP;
+	}
+
+	return cell->dev->nvmem_ops->read_cell(cell, output_buf, buf_len,
+					       read_len);
+}
+
+int nvmem_cell_write(struct nvmem_cell *cell, uint8_t *input_buf,
+		     size_t buf_len)
+{
+	if (cell->dev->nvmem_ops->write_cell == NULL) {
+		return -ENOTSUP;
+	}
+
+	return cell->dev->nvmem_ops->write_cell(cell, input_buf, buf_len);
+}
diff --git a/include/drivers/st/nvmem.h b/include/drivers/st/nvmem.h
new file mode 100644
index 0000000000..0bc1a421a6
--- /dev/null
+++ b/include/drivers/st/nvmem.h
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2023, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef NVMEM_H
+#define NVMEM_H
+
+#include <stddef.h>
+#include <stdint.h>
+
+struct nvmem_dev {
+	const struct nvmem_ops *nvmem_ops;
+	void *drv_data;
+	int offset;
+};
+
+struct nvmem_cell {
+	int offset;
+	size_t len;
+	struct nvmem_dev *dev;
+};
+
+struct nvmem_ops {
+	int (*read_cell)(struct nvmem_cell *cell, uint8_t *data, size_t buf_len,
+			 size_t *read_len);
+	int (*write_cell)(struct nvmem_cell *cell, uint8_t *data, size_t len);
+};
+
+int nvmem_driver_register(int nodeoffset, const struct nvmem_ops *nvmem_ops,
+			  void *drv_data);
+
+int nvmem_get_cell_by_index(const void *fdt, int nodeoffset, unsigned int index,
+			    struct nvmem_cell *cell);
+
+int nvmem_get_cell_by_name(const void *fdt, int nodeoffset, const char *name,
+			   struct nvmem_cell *cell);
+
+int nvmem_cell_read(struct nvmem_cell *cell, uint8_t *output_buf,
+		    size_t buf_len, size_t *read_len);
+
+int nvmem_cell_write(struct nvmem_cell *cell, uint8_t *input_buf,
+		     size_t buf_len);
+#endif
diff --git a/plat/st/stm32mp1/stm32mp1_def.h b/plat/st/stm32mp1/stm32mp1_def.h
index f8742106b7..4d3fb1dc68 100644
--- a/plat/st/stm32mp1/stm32mp1_def.h
+++ b/plat/st/stm32mp1/stm32mp1_def.h
@@ -697,6 +697,9 @@ static inline uintptr_t tamp_bkpr(uint32_t idx)
 #define PLAT_NB_FIXED_REGUS		U(2)
 /* No GPIO regu */
 #define PLAT_NB_GPIO_REGUS		U(0)
+/* TAMP BACKUP REGISTER */
+#define PLAT_NB_NVMEM_DEVS		U(1)
+#define MAX_TAMP_BACKUP_REGS_ZONES	U(3)
 
 /*******************************************************************************
  * Device Tree defines
diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index b84627bcf8..3f325769bb 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -561,6 +561,9 @@ static inline uintptr_t tamp_bkpr(uint32_t idx)
 #define PLAT_NB_GPIO_REGUS			U(0)
 /* Number of low power modes defined in the device tree */
 #define PLAT_NB_SUSPEND_MODES		7
+/* TAMP BACKUP REGISTER */
+#define PLAT_NB_NVMEM_DEVS		U(1)
+#define MAX_TAMP_BACKUP_REGS_ZONES	U(7)
 
 /*******************************************************************************
  * Device Tree defines
-- 
2.53.0

